### PR TITLE
Cleanup selector usage and refs

### PR DESCRIPTION
**Pull request by rgbkrk**
Doing a sweep on the selectors to get rid of improper memoization with createSelector and enforce more types within use of selectors.

### CODE REVIEW

File: packages/core/src/selectors/index.js [hunk 1]
```diff
@@ -3,13 +3,17 @@
...
+  JupyterHostRecord,
+  CoreRecord
+} from "../state";
 
 // FIXME FIXME FIXME SUPER WRONG FIXME FIXME FIXME
 type AppState = {
   // The new way
-  core: any,
-
+  core: CoreRecord,
```
Comment (theengineear) [inline packages/core/src/selectors/index.js:19]: ✨

**Comment by rgbkrk** [comment]
Jupyter extension is working well

Desktop needs some work due to me locking down the contentRefs on the notebook app.

File: packages/core/src/selectors/index.js [hunk 2]
```diff
@@ -53,29 +57,39 @@ export const currentHost = createSelector(
   identity
 );
 
-export const kernelsByRef = createSelector(
-  (state: AppState) => state.core.getIn(["entities", "kernels", "byRef"]),
-  identity
```
Comment (theengineear) [inline packages/core/src/selectors/index.js:58]: Nice. Yah, this definitely didn't make sense before.

**Comment by rgbkrk** [comment]
Desktop all set now, though we certainly want to clean up how we render contents for the placeholder gunk.

File: applications/notebook-on-next/store.js [hunk 3]
```diff
@@ -1,4 +1,5 @@
-/* @flow */
+// FIXME: To make it easier to iterate on supported apps, flow is disabled for
+// this file
```
Comment (theengineear) [inline applications/notebook-on-next/store.js:2]: 👌

File: packages/core/src/selectors/notebook.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
+// @flow
+import * as Immutable from "immutable";
+
+import * as commutable from "@nteract/commutable";
+import type { CellID } from "@nteract/commutable";
+
+import { createSelector } from "reselect";
+
+// All these selectors expect a NotebookModel as the top level state
```
Comment (theengineear) [inline packages/core/src/selectors/notebook.js:9]: Nice. I like that these are split out now 👍

File: packages/core/src/selectors/notebook.js [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+export const editorFocusedId = (model: NotebookModel) => model.editorFocused;
+
+export const codeCellIdsBelow = createSelector(
+  [cellMap, cellOrder, cellFocused],
+  (cellMap, cellOrder, cellFocused) => {
+    if (!cellFocused) {
+      return Immutable.List();
+    }
+    const index = cellOrder.indexOf(cellFocused);
+    // NOTE: if there is no focused cell, this runs all of them
```
Comment (theengineear) [inline packages/core/src/selectors/notebook.js]: ~runs~ returns

Comment (theengineear) [reply packages/core/src/selectors/notebook.js]: Wait... this isn't true. You're returning an empty list above aren't you? What's the intended behavior?

File: packages/core/src/selectors/notebook.js [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+);
+
+export const gistId = createSelector([metadata], metadata =>
+  metadata.get("gist_id", null)
+);
+
+export const notebook = (model: NotebookModel) => model.notebook;
+export const savedNotebook = (model: NotebookModel) => model.savedNotebook;
+
+export const hasBeenSaved = createSelector(
```
Comment (theengineear) [inline packages/core/src/selectors/notebook.js]: nit, `hasBeenSaved` would be like `!!disk` and `isDirty` would be like `Immutable.is(original, disk)`, right? Just a little confused by the naming?

File: packages/core/src/selectors/notebook.js [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  notebook,
+  savedNotebook,
+  (original, disk) => Immutable.is(original, disk)
+);
+
+export const asJSON = createSelector([notebook], notebook => {
+  return commutable.toJS(notebook);
+});
```
Comment (theengineear) [inline packages/core/src/selectors/notebook.js:95]: 💥

File: packages/core/src/selectors/notebook.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+// NOTE: This is called asString instead of toString so that REPLs
```
Comment (theengineear) [inline packages/core/src/selectors/notebook.js:97]: 👌

File: packages/core/src/components/notebook-menu/index.js [hunk 9]
```diff
@@ -401,12 +401,26 @@ class PureNotebookMenu extends React.Component<Props, State> {
...
-  currentKernelRef: selectors.currentKernelRef(state),
-  currentContentRef: selectors.currentContentRef(state)
-});
+const mapStateToProps = state => {
+  const contentRef = selectors.currentContentRef(state);
+  if (!contentRef) {
+    throw new Error("There must be a contentRef for this menu");
+  }
+
+  // TODO: Move the downloader to an epic so we're not putting the notebook _in_
```
Comment (theengineear) [inline packages/core/src/components/notebook-menu/index.js:410]: 👍

**Review by theengineear — COMMENTED** [review]
Looks great, Kyle! I just have a couple comments in here for ya!

File: packages/core/src/selectors/notebook.js (hunk 6 above)
Comment (rgbkrk) [reply packages/core/src/selectors/notebook.js]: I like isDirty since that's what we end up using this for, I'll switch that up.

Comment (rgbkrk) [reply packages/core/src/selectors/notebook.js]: Done!

File: packages/core/src/selectors/index.js (hunk 1 above)
Comment (rgbkrk) [reply packages/core/src/selectors/index.js:19]: 😃

**Pull request merged**